### DISCUSSION TITLE
What about "this" where clause for generic interfaces

### DISCUSSION THREAD

**Discussion by Matt90hz** [dotnet/csharplang#7270] [Language Ideas] [answered]
I came across this while trying to implement a service that works like a post office. So we have a sender, a receiver and a service that transports information between the two. Classic mediator pattern.

One of the feature that I wanted is that a sender must have a receiver on the other end. So I implemented interfaces like:

```Csharp
public interface ISender<TSender, TReceiver, TState>
    where TReceiver : IReceiver<TReceiver, TSender, TState>
    where TSender : ISender<TSender, TReceiver, TState>
{

}

public interface IReceiver<TReceiver , TSender, TState> 
    where TSender : ISender<TSender, TReceiver, TState>
    where TReceiver : IReceiver<TReceiver, TSender, TState>
{

}
```
This works great and if there are no matching signatures for the generic parameters the compiler signals an error. The only thing that bugged me is that any type can implement such interfaces while I would have liked to be able to impose that only the class that implements this interfaces is eligible to be the sender or the receiver. So I would like that C# had a where clause like "self", "me" or "this" to bind a generic type to the type that implements the interface. Like:

```Csharp
public interface ISender<TSender, TReceiver, TState>
    where TReceiver : IReceiver<TReceiver, TSender, TState>
    where TSender : this, ISender<TSender, TReceiver, TState>
{

}

public interface IReceiver<TReceiver , TSender, TState> 
    where TSender : ISender<TSender, TReceiver, TState>
    where TReceiver : this, IReceiver<TReceiver, TSender, TState
{

}
```
In this way declaring something like this would be not allowed:

```Csharp
class MyClass : ISender<OtherClass, ReceiverClass, int> //Compiler error
{

}

class MyClass : ISender<MyClass, ReceiverClass, int> //Compiler OK
{

}
```
As of today I did not find any simple way around this. Of course I could throw an exception at runtime but is not a solution. The only way that I think it might work is build an analyzer but is far from my definition of simple solution.

I also would find this feature syntactic sugar if the clause could be specified in the interface declaration as for the extension methods, so during the implementation can be omitted. Like this:

```Csharp
public interface ISender<this, TReceiver, TState>
    where TReceiver : IReceiver<this, TState>
{

}

public interface IReceiver<this, TSender, TState> 
    where TSender : ISender<this, TState>
{

}

class SenderClass : ISender<ReceiverClass, int> 
{

}

class ReceiverClass : IReceiver<SenderClass, int>
{

}
```

Maybe is not the most useful feature but I think it has potential in many scenarios beyond the one that I came across.

**Comment by 333fred** [comment, marked as answer]
This is most similar to https://github.com/dotnet/csharplang/issues/5413#issuecomment-1081313483, which was rejected by the LDM. We are interested in self types, just not this version of them.

**Reply by Matt90hz** [reply]
Yeah the same concept. 
Sorry I did not found this discussion, my bad.
Well I hope it will be implemented sooner or later.
Thanks.
